### PR TITLE
(COV-816): refactor/chartTotalAdjustment: have charts use global total

### PR DESCRIPTION
**Pull request by ocshawn**
Jira Ticket: [COV-816](https://occ-data.atlassian.net/browse/COV-816)

### Breaking Changes
- SummaryChartGroup requires totalCount property 
- SummaryPieChart requires totalCount property 
- SummaryBarChart requires totalCount property 

### Improvements
- have charts use global total not total of data in chart

### CODE REVIEW

File: src/components/charts/SummaryChartGroup/index.jsx [hunk 1]
```diff
@@ -80,6 +83,7 @@ class SummaryChartGroup extends Component {
 
 SummaryChartGroup.propTypes = {
   summaries: PropTypes.arrayOf(PropTypes.object).isRequired,
+  totalCount: PropTypes.number.isRequired,
```
Comment (mfshao) [inline src/components/charts/SummaryChartGroup/index.jsx]: I'm in favor of making this new props an optional one, if omitted, use old logic to calculate percentage

File: stories/charts.js [hunk 2]
```diff
@@ -102,34 +103,34 @@ const customizedColorMap = [
...
-    <SummaryChartGroup summaries={summaries} width={1010} />
+    <SummaryChartGroup summaries={summaries} width={1010} totalCount={10000} />
   ))
   .add('SummaryChartGroup with only showing 2 at first', () => (
-    <SummaryChartGroup summaries={summaries} width={1010} maximumDisplayItem={2} />
+    <SummaryChartGroup summaries={summaries} width={1010} maximumDisplayItem={2} totalCount={1000} />
   ))
   .add('SummaryChartGroup with a Locked Chart', () => (
-    <SummaryChartGroup summaries={lockedSummaries} width={1010} lockValue={-1} lockMessage='This chart is locked!' />
+    <SummaryChartGroup summaries={lockedSummaries} width={1010} lockValue={-1} lockMessage='This chart is locked!' totalCount={1000} />
```
Comment (mfshao) [inline stories/charts.js]: This logic needs be able to handle lock values `-1` by default, we cannot hard code this total count if the data returned has lock values

File: stories/charts.js [hunk 3]
```diff
@@ -102,34 +103,34 @@ const customizedColorMap = [
...
-    <SummaryHorizontalBarChart data={virusData} title='bar chart title' showPercentage={false} />
+    <SummaryHorizontalBarChart data={virusData} title='bar chart title' showPercentage={false} totalCount={virusDataTotal}/>
   ))
   .add('SummaryHorizontalBarChart with single color and percentage', () => (
-    <SummaryHorizontalBarChart data={virusData} title='bar chart title' color='#3283c8' />
+    <SummaryHorizontalBarChart data={virusData} title='bar chart title' color='#3283c8'  totalCount={virusDataTotal} />
   ))
   .add('SummaryHorizontalBarChart with customized colors', () => (
-    <SummaryHorizontalBarChart data={virusData} title='bar chart title' showPercentage={false} useCustomizedColorMap customizedColorMap={customizedColorMap} />
+    <SummaryHorizontalBarChart data={virusData} title='bar chart title' showPercentage={false} useCustomizedColorMap customizedColorMap={customizedColorMap}  totalCount={virusDataTotal}/>
```
Comment (mfshao) [inline stories/charts.js]: there are multiple eslint issues in this file

**Review by mfshao — CHANGES_REQUESTED** [review]
I do think we need to re-visit the design in the ticket. IMO it is even more confusing when user see something like this
Screen Shot 2021-04-02 at 6 29 11 PM

How come there could be a full pie chart but the percentages doesn't add up to `100%`?

If we want to do this, better fill-in the missing counts with `no data` or `N/A`, so the chart can has that info displayed and total percentage adds up to `100%`

File: stories/charts.js (hunk 2 above)
Comment (ocshawn) [reply stories/charts.js]: normally in the data portal this value will be calculated automatically based on props.totalCount but i will remove all of the hard coded values and implement the old logic if the totalCount is omitted as you suggested

Comment (mfshao) [reply stories/charts.js]: Oh yeah I get it, thanks. 

But I'd suggest let's revisit the design in this ticket with its owner, before you spending some effort in here. I just don't want them to be wasted

File: stories/charts.js (hunk 3 above)
Comment (ocshawn) [reply stories/charts.js]: dont know why my linter does not lint this file, but i ran it manually

**Comment by ocshawn** [comment]
Reached out to the ticket holder and they are fine with a “No Data” field that fills in the missing data. This has been updated along with running eslint on the charts.js and adding some more stories to make sure everything looks correct.

**Comment by mfshao** [comment]
> Reached out to the ticket holder and they are fine with a “No Data” field that fills in the missing data. This has been updated along with running eslint on the charts.js and adding some more stories to make sure everything looks correct.

I have an impression that what you are trying to do here is already supported in Guppy. Have you tried to turn on `aggs_include_missing_data` for Guppy in the manifest? https://github.com/uc-cdis/cdis-manifest/blob/master/chicagoland.pandemicresponsecommons.org/manifest.json#L237

If this option is on (it should be on by default, I don't know why it is turned off for PRC, you should check with the team), Guppy should include counts for missing values in the returned results for aggregation queries. See https://github.com/uc-cdis/guppy#configurations

One thing you should notice is that by turning this option on, there might be some cases for an unauthroized user that can never see some charts if there is a small fraction of data that don't have some values. But that is an expected behavior, and whatever you are doing here should also honors that logic. That is, if the aggregation result contains a lock value, then don't display the chart. I don't think that logic is addressed in this PR anyways.

File: src/components/charts/SummaryPieChart/index.jsx [hunk 4]
```diff
@@ -32,16 +32,33 @@ class SummaryPieChart extends React.Component {
...
+    // if chart is missing total add it
+    const dataSum = tempDataArr.reduce((a, entry) => a + entry.value, 0);
+    if (!tempDataTotal) {
+      tempDataTotal = dataSum;
+
+    // if chart is missing data add it
+    } else if (tempDataTotal > dataSum) {
+      tempDataArr.push({
+        name: 'No Data',
+        value: tempDataTotal - dataSum,
```
Comment (mfshao) [inline src/components/charts/SummaryPieChart/index.jsx:46]: possible info leak in here: if env is in `regular` tier access level, this value may be less than the `tiered_access_limit` value, and should cause the chart get locked

**Comment by ocshawn** [comment]
taking completely different approach see https://github.com/uc-cdis/guppy/pull/120

**Pull request closed without merging**